### PR TITLE
Initial implementations of student page.

### PR DESCRIPTION
**Pull request by abeeto**
Requirements:
- search bar to search by student (last name for now)
Work done so far:
- set of sections parsed from table ( useEffect in page.tsx)
- table has two columns for name: firstname and lastname, it parses the student name attribute (by splitting on ",")
- section filtering can be done on client side

UPDATE Feb 25, 2024:
- used customHooks and context as mentioned in meeting to clean up code and avoid prop drilling.
- refactor: deviated from the generic table we had to follow the structure/layout as seen in ShadCN's tasks table example. 
- new components for the data table are under the students folder page.
- sorting and filtering sections now works

UPDATE Mar 04, 2024:
- removed teamsets filter 

**custom components:**
-  students/components/students-table.tsx
-  students/components/data-table-upload-dialog.tsx

**shadcn components**
- students/components/data-table-(faceted-filter/pagination/toolbar/view-options) were all taken out of shadcn Tasks [example](https://github.com/shadcn-ui/ui/tree/main/apps/www/app/examples/tasks/components).

- To note, in general ShadCN has no built in "plug and play" data-table component, they seemed to build it on top of tanstack tables. Leaving these inside students folder for now as they need to be consolidated.

### CODE REVIEW

File: src/app/students/hooks/useTeamSets.tsx [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+    const [displayTeamSet, setDisplayTeamSet] = useState<TeamSet>()
+
+    useEffect(() => {
+        // Assume getTeamSets() is a function that fetches team sets
+        const teamSets = getTeamSets() // Adjust this call to your actual data fetching logic
+        setAllTeamSets(teamSets)
+    }, [])
+
+    const handleSelectTeamSet = (teamSetIdString: string) => {
+        const teamSetId = parseInt(teamSetIdString)
```
Comment (Opeyem1a) [inline src/app/students/hooks/useTeamSets.tsx:36]: wait why is it ever a string if we just parse it to an int anyway

File: src/app/students/hooks/useTeamSets.tsx [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+
+export const TeamSetsProvider: React.FC<PropsWithChildren> = ({ children }) => {
+    const teamSets = useTeamSetsProvider()
+    return <TeamSetsContext.Provider value={teamSets}>{children}</TeamSetsContext.Provider>
+}
+
+export const useTeamSets = (): TeamSetsContextType => {
+    const context = useContext(TeamSetsContext)
+    if (context === undefined) {
+        throw new Error('useTeamSets must be used within a TeamSetsProvider')
```
Comment (Opeyem1a) [inline src/app/students/hooks/useTeamSets.tsx:56]: This error gets thrown by default and the condition you are checking does not prove this is the correct error. I would just remove this whole if block

File: src/app/students/views/page.tsx
Comment (Opeyem1a) [inline src/app/students/views/page.tsx:1]: why is this file entirely new? Shouldn't it just have updated from what is in main?

Either you've moved this file somehow or something funky is happening with your git

File: src/app/students/views/page.tsx [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+    breadcrumbs?: Array<{
+        title: string;
+        href: string;
+    }>
+};
+
+const PageView = ({
+    children,
+    title,
+    rightSide,
```
Comment (Opeyem1a) [inline src/app/students/views/page.tsx:17]: nit;

rightSide is a bad name imo

File: src/app/students/views/studentsPage.tsx
Comment (Opeyem1a) [inline src/app/students/views/studentsPage.tsx:1]: this is not the intended use of the `/views`

Views in this repo should represent different structures of a page. So far, we really only have the PageView and a maybe a SplitView (not yet coded) in the designs 

If this is specific to the students page, you could define a sublayout in the students directory

File: src/app/students/views/studentsPage.tsx [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+    })
+    return students
+}
+
+
+export default function StudentsPage() {
+    const { allTeamSets, displayTeamSet, handleSelectTeamSet } = useTeamSets()
+    const { displayStudents} = useStudents()
+
+    const TeamSelect = () => {
```
Comment (Opeyem1a) [inline src/app/students/views/studentsPage.tsx:36]: Don't define components inside of other components, this is what we talked about in our meeting

File: src/app/students/views/studentsPage.tsx [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+                {<SelectContent>
+                    {allTeamSets.map((teamSet) => (
+                        <SelectItem value={"" + teamSet.id} key={teamSet.id}>{teamSet.name}</SelectItem>
+                    ))}
+                </SelectContent>}
+            </Select>
+        )
+    }
+    return (
+        <PageView title="Students" rightSide={<TeamSelect/>} breadcrumbs={[
```
Comment (Opeyem1a) [inline src/app/students/views/studentsPage.tsx:52]: I don't really know if having the team select here is worth making this whole page 'use client'. This TeamSelect thing can just be defined as it's own component file that is 'use client'

File: src/app/students/views/studentsPage.tsx [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+import PageView from '@/app/students/views/page'
+// import {DataTable} from '@/components/ui/data-table'
+import {Student} from '@/_temp_types/student'
+import {Select, SelectContent, SelectItem, SelectTrigger, SelectValue} from "@/components/ui/select"
+import { useStudents } from '../hooks/useStudents'
+import { useTeamSets } from '../hooks/useTeamSets'
+import {columns} from '../columns'
+import {DataTable} from "../components/data-table"
+
+function getStudentData() {
```
Comment (Opeyem1a) [inline src/app/students/views/studentsPage.tsx:12]: is this function even currently being used? Don't leave dead code in PRs

File: src/app/students/hooks/useStudents.tsx [hunk 7]
```diff
@@ -0,0 +1,72 @@
+import { useState, useEffect, useCallback, createContext, useContext, ChangeEvent, PropsWithChildren } from 'react'
+import { parseCSV } from '@/lib/parseCSV'
+import { Student } from '@/_temp_types/student'
+
+interface SectionOption {
```
Comment (Opeyem1a) [inline src/app/students/hooks/useStudents.tsx:5]: Same note about using types unless you need an interface's functionality

File: src/app/students/hooks/useStudents.tsx [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+  currentSections: SectionOption[];
+  handleFileChange: (event: ChangeEvent<HTMLInputElement>) => Promise<void>;
+  handleCancel: () => void;
+  handleSave: () => void;
+}
+
+const StudentsContext = createContext<StudentsContextType | undefined>(undefined)
+
+export const useStudentsProvider = (): StudentsContextType => {
+    const [csvStudentsParse, setStudentsParse] = useState<Student[]>([])
```
Comment (Opeyem1a) [inline src/app/students/hooks/useStudents.tsx:21]: Follow convention here, should always be `useA, setA`

File: src/app/students/hooks/useStudents.tsx [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+        displayStudents,
+        currentSections,
+        handleFileChange,
+        handleCancel,
+        handleSave,
+    }
+}
+
+export const StudentsProvider: React.FC<PropsWithChildren> = ({ children }) => {
+    const students = useStudentsProvider()
```
Comment (Opeyem1a) [inline src/app/students/hooks/useStudents.tsx:62]: this custom hook isn't needed either, and if it was, `students` is a mega misleading name

File: src/app/students/hooks/useStudents.tsx [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+    return <StudentsContext.Provider value={students}>{children}</StudentsContext.Provider>
+}
+
+export const useStudents = (): StudentsContextType => {
+    const context = useContext(StudentsContext)
+    if (context === undefined) {
+        throw new Error('useStudents must be used within a StudentsProvider')
```
Comment (Opeyem1a) [inline src/app/students/hooks/useStudents.tsx:69]: Same comment as the other context

File: src/app/students/hooks/useStudents.tsx [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+
+interface SectionOption {
+  label: string;
+  value: string;
+}
+
+interface StudentsContextType {
+  displayStudents: Student[];
+  currentSections: SectionOption[];
+  handleFileChange: (event: ChangeEvent<HTMLInputElement>) => Promise<void>;
```
Comment (Opeyem1a) [inline src/app/students/hooks/useStudents.tsx:13]: Idk if this is needed, my understanding was that once they load students in, they would be saved to the DB, thus there's no such thing as changing the file, only adding new files (which would perform an upsert in the backend)

File: src/app/students/hooks/useStudents.tsx [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+}
+
+const StudentsContext = createContext<StudentsContextType | undefined>(undefined)
+
+export const useStudentsProvider = (): StudentsContextType => {
+    const [csvStudentsParse, setStudentsParse] = useState<Student[]>([])
+    const [displayStudents, setDisplayStudents] = useState<Student[]>([])
+    const [currentSections, setSections] = useState<SectionOption[]>([])
+
+    useEffect(() => {
```
Comment (Opeyem1a) [inline src/app/students/hooks/useStudents.tsx:25]: This useEffect should have been a useMemo

File: src/app/students/hooks/useStudents.tsx
Comment (Opeyem1a) [inline src/app/students/hooks/useStudents.tsx:1]: It would be wise to remember that most of the time, the data for students is coming from the backend, and being able to load students through a CSV is basically just a convenient shortcut to manually filling out a "create student" form 200 times

I should have mentioned this in the design phase, but the CSV thing shouldn't be this front and center, it's really just one of (hopefully) many options to add students to your course

**Review by Opeyem1a — CHANGES_REQUESTED** [review]
Didn't get a chance to review your custom data table components but reckoned there iss more than enough work here to do already

I don't want you to be discouraged, I think the real underlying issue here is an incomplete design phase, not even time assigned to cleaning up your PR, and general confusion in my instructions and feedback.

I would encourage you to ask me for more context if you feel something was vague instead of trying to hack together a vague understanding from guides. Following tutorials on your own is good, but when you're implementing like 4-5 brand new concepts to you and all of them are done this way, you're bound to misstep a lot.

Try to work through these comments and remember, all of the things I said are just suggestions, the goal is not to implement each new thing I mention, but to consider the project's needs and assess how this new tool can help you accomplish them.


Also, lint the code, it will help with the formatting bugs

File: src/_temp_types/student.ts [hunk 13]
```diff
@@ -2,9 +2,11 @@ import {Relationship} from "@/_temp_types/relationship"
 
 export type Student = {
     id: number
-    name?: string
+    name: string,
```
[the remark this answers is not in the document]
Comment (abeeto) [reply src/_temp_types/student.ts:5]: I thought canvas wouldn't accept (student objects) without a name.

File: src/_temp_types/teamSet.ts [hunk 14]
```diff
@@ -2,6 +2,6 @@ import {Team} from "@/_temp_types/team"
 
 export type TeamSet = {
     id: number
-    name?: string
+    name: string
```
[the remark this answers is not in the document]
Comment (abeeto) [reply src/_temp_types/teamSet.ts:5]: I think this change was done by Kiet during our peer code session. Changing it back to optional for now.

File: src/app/students/columns.tsx [hunk 15]
```diff
@@ -0,0 +1,50 @@
...
+            ))
+            return <div className="flex flex-row gap-2">
+                {sectionBadges}
+            </div>
+        },
+        filterFn: (row, id, filterValues: SectionFilterValue) => {
+            // Get the row's sections as an array
+            const rowSections = row.getValue(id) as string[]
+            // Check if any of the filterValues is included in the row's sections
+            return filterValues.every(filterValue => rowSections.includes(filterValue))
```
[the remark this answers is not in the document]
Comment (abeeto) [reply src/app/students/columns.tsx:46]: I think I misunderstood the intended behavior it's filtering each row of the table on the condition that the row must have each and every filter (section) selected by the user. But thinking about it that doesn't make much sense for this context. Changed .every to .some? This way it checks if Student is any of the selected sections not if they are in every selected section.
```suggestion
    return filterValues.some(filterValue => rowSections.includes(filterValue))
```

File: src/app/students/columns.tsx [hunk 16]
```diff
@@ -0,0 +1,50 @@
...
+import { Badge } from "@/components/ui/badge"
+import { DataTableColumnHeader } from "@/components/ui/table-column-header"
+
+type SectionFilterValue = string[]; // Define the type for the filter value
+
+export const columns: ColumnDef<Student>[] = [
+    {
+        id: "firstName",
+        header: ({ column }) => <DataTableColumnHeader column={column} title="First Name" hasDropDownMenu={false}/>,
+        accessorFn: (row) => row.name.split(',')[1], // Assuming the first part is the first name
```
Comment (abeeto) [inline src/app/students/columns.tsx:14]: eslint giving " 'row.name' is possibly 'undefined'.ts(18048) " as a result of making the name field of Student as optional.

File: src/app/students/page.tsx [hunk 17]
```diff
@@ -0,0 +1,16 @@
+// App.js or a parent component
+'use client'
```
[the remark this answers is not in the document]
Comment (abeeto) [reply src/app/students/page.tsx:1]: Seem to need this because if I remove it I get an error about useStudents not being able to be resolved...
Tried to move 'use client' to useStudents where StudentsProvider is but that doesn't seem to work.

File: src/components/ui/command.tsx
[the remark this answers is not in the document]
Comment (abeeto) [reply src/components/ui/command.tsx:1]: Understood. Will update PR description to include the imports from shadcn. The command component was installed from shadcn component, was always there.

File: src/components/views/Page.tsx [hunk 18]
```diff
@@ -21,12 +21,12 @@ const PageView = ({
                 {breadcrumbs && (
                     <div className="flex gap-1">
                         {breadcrumbs.map((breadcrumb, index) => (
-                            <>
-                                <Link key={index} href={breadcrumb.href}>
+                            <div key = {index} className="flex flex-row">
```
Comment (abeeto) [inline src/components/views/Page.tsx:24]: made this change as I was getting an error about not having a key prop for each child

File: src/app/students/(table)/columns.tsx [hunk 19]
```diff
@@ -0,0 +1,46 @@
...
+import { DataTableColumnHeader } from "@/components/ui/table-column-header"
+
+type SectionFilterValue = string[];
+
+export const columns: ColumnDef<Student>[] = [
+    {
+        id: "firstName",
+        header: ({ column }) => <DataTableColumnHeader column={column} title="First Name" hasDropDownMenu={false}/>,
+        accessorFn: (row) => row.name.split(',')[1],
+        cell: ({ getValue }) => <div className="text-left font-sm">{String(getValue())}</div>,
```
Comment (Opeyem1a) [inline src/app/students/(table)/columns.tsx:15]: complete nitpick, do as a separate PR, but these divs should probably be `<Text />` components

File: src/app/students/page.tsx [hunk 20]
```diff
@@ -0,0 +1,19 @@
+// 'use client'
```
Comment (Opeyem1a) [inline src/app/students/page.tsx:1]: nit:

dead comment

**Review by Opeyem1a — APPROVED** [review]
Looks good, note that if I approve a PR, that means you can merge it, and any comments I leave on an approved PR are nitpicks and/or "you can merge this now and do that later" comments
